File: dart/servewright_flutter/lib/src/renderer.dart
import 'package:flutter/widgets.dart';
import 'package:servewright_flutter/src/types.dart';

class RenderContext {
  const RenderContext({required this.renderChild});

  final Widget Function(ServewrightNode node) renderChild;

  List<Widget> renderChildren(List<ServewrightNode> nodes) =>
      nodes.map(renderChild).toList();
}

typedef PrimitiveBuilder = Widget Function(ServewrightNode node, RenderContext ctx);

class Registry {
  final Map<String, PrimitiveBuilder> _builders = {};

  void register(String type, PrimitiveBuilder builder) {
    _builders[type] = builder;
  }

  PrimitiveBuilder? resolve(String type) => _builders[type];
}

class Renderer {
  Renderer(this.registry);

  final Registry registry;

  Widget render(ServewrightView view) => _renderNode(view.root);

  Widget _renderNode(ServewrightNode node) {
    final builder = registry.resolve(node.type);
    if (builder == null) {
      return _unknownPlaceholder(node);
    }
    return builder(node, RenderContext(renderChild: _renderNode));
  }

  Widget _unknownPlaceholder(ServewrightNode node) {
    return SizedBox(
      key: ValueKey('unknown-${node.id}'),
      child: Text('Unknown primitive: ${node.type}'),
    );
  }
}

Registry createRegistry() => Registry();

Renderer createRenderer(Registry registry) => Renderer(registry);

File: js/servewright-react/src/renderer.ts
import { createElement, type ReactElement } from "react";
import type { ServewrightNode, View } from "./types.js";

export type PrimitiveComponent = (node: ServewrightNode) => ReactElement;

export interface Registry {
  register(type: string, component: PrimitiveComponent): void;
  resolve(type: string): PrimitiveComponent | undefined;
}

export function createRegistry(): Registry {
  const components = new Map<string, PrimitiveComponent>();

  return {
    register(type, component) {
      components.set(type, component);
    },
    resolve(type) {
      return components.get(type);
    },
  };
}

export interface Renderer {
  render(view: View): ReactElement;
}

export function createRenderer(registry: Registry): Renderer {
  const renderNode = (node: ServewrightNode): ReactElement => {
    const component = registry.resolve(node.type);
    if (!component) {
      return createUnknownPlaceholder(node);
    }
    return component(node);
  };

  return {
    render(view) {
      return renderNode(view.root);
    },
  };
}

function createUnknownPlaceholder(node: ServewrightNode): ReactElement {
  return createElement(
    "div",
    {
      "data-servewright-unknown": node.type,
      "data-servewright-id": node.id,
    },
    `Unknown primitive: ${node.type}`,
  );
}
